cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -268,78 +282,370 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+	// "fully" degraded mode, but we do need to track that we had to
+	// use the fallback key
+	ctx.degradedState.DataKey = "fallback"
+	ctx.dataDevice = dataDevice
+
+	return m.mountData, nil
+}
+
+func (m *stateMachine) mountData(ctx *context) (stateFunc, error) {
 	// don't do fsck on the data partition, it could be corrupted
-	if err := doSystemdMount(device, boot.InitramfsHostUbuntuDataDir, nil); err != nil {
-		return err
+	if err := doSystemdMount(ctx.dataDevice, boot.InitramfsHostUbuntuDataDir, nil); err != nil {
+		// we failed to mount it, proceed with degraded mode
+		ctx.degradedState.DataState = "not-mounted"
+
+		// no point trying to unlock save with the run key, we need data to be
+		// mounted for that and we failed to mount it
+		return m.unlockSaveFallbackKey, nil
+	}
+	// we mounted it successfully, verify it comes from the right disk
+	if err := verifyMountPointCtx(ctx, boot.InitramfsHostUbuntuDataDir, "ubuntu-data"); err != nil {
+		return nil, err
 	}
 
-	// 3.1. mount ubuntu-save (if present)
-	haveSave, err := maybeMountSave(disk, boot.InitramfsHostWritableDir, isDecryptDev, nil)
+	ctx.degradedState.DataState = "mounted"
+	ctx.degradedState.HostLocation = boot.InitramfsHostUbuntuDataDir
+
+	// next step: try to unlock with run save key
+	return m.unlockSaveRunKey, nil
+}
+
+func (m *stateMachine) locateUnencryptedSave(ctx *context) (stateFunc, error) {
+	partUUID, err := ctx.disk.FindMatchingPartitionUUID("ubuntu-save")
 	if err != nil {
-		return err
+		// error locating ubuntu-save
+		if _, ok := err.(disks.FilesystemLabelNotFoundError); ok {
+			// this is ok, ubuntu-save may not exist for
+			// non-encrypted device
+			ctx.degradedState.SaveState = "not-needed"
+		} else {
+			// the error is not "not-found", so we have a real error
+			// identifying whether save exists or not
+			logger.Noticef("error identifying ubuntu-save partition: %v", err)
+			ctx.degradedState.SaveState = "other"
+		}
+
+		// all done, nothing left to try and mount
+		return nil, errStateDone
 	}
 
-	// 3.2 verify that the host ubuntu-data comes from where we expect it to
-	// right device
-	diskOpts := &disks.Options{}
-	if isDecryptDev {
-		// then we need to specify that the data mountpoint is expected to be a
-		// decrypted device, applies to both ubuntu-data and ubuntu-save
-		diskOpts.IsDecryptedDevice = true
+	// we found the unencrypted device, now mount it
+	ctx.saveDevice = filepath.Join("/dev/disk/by-partuuid", partUUID)
+	return m.mountSave, nil
+}
+
+func (m *stateMachine) unlockSaveRunKey(ctx *context) (stateFunc, error) {
+	// XXX: would be nice to not need this redirect here
+	if !ctx.isEncryptedDev {
+		return m.locateUnencryptedSave, nil
+	}
+
+	// to get to this state, we needed to have mounted ubuntu-data on host, so
+	// if encrypted, we can try to read the run key from host ubuntu-data
+	saveKey := filepath.Join(dirs.SnapFDEDirUnder(boot.InitramfsHostWritableDir), "ubuntu-save.key")
```

Comment:
Now that the data mountpoint is passed in context, this could add a helper that returns the patht o writable under given data mount rootdir, eg. `dirs.SnapFDEDirUnder(boot.HostWritableDirUnder(ctx.ctx.degradedState.HostLocation))`, wdyt?